[Text Extraction] Retrieved items should be reported in the coordinate space of the web view

whsieh opened the pull request:
#### 884b0d2472da61ab72615727088039ab05e998a6
<pre>
[Text Extraction] Retrieved items should be reported in the coordinate space of the web view
<a href="https://bugs.webkit.org/show_bug.cgi?id=270994">https://bugs.webkit.org/show_bug.cgi?id=270994</a>
<a href="https://rdar.apple.com/124557587">rdar://124557587</a>

Reviewed by Abrar Rahman Protyasha.

`WKTextExtractionItem`&apos;s rect should be in web view coordinates, rather than root view coordinates.
Fix this by refactoring the text extraction item conversion code, so that we can pass in a function
to convert from root view coordinates to web view coordinates if needed (i.e. on iOS).

Test: fast/text-extraction/text-extraction-when-scrolled.html

* LayoutTests/fast/text-extraction/text-extraction-when-scrolled-expected.txt: Added.
* LayoutTests/fast/text-extraction/text-extraction-when-scrolled.html: Added.

Add a new layout test to exercise the change:
1.  Only the bottom link should be extracted
2.  The y-offset of the extracted link should be closer to 0 than 10000 (roughly the full height of
    the view).
3.  The y-offset of the root should be a large negative number close to 10000.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestTextExtraction):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm:
(WebKit::createWKTextItem):
(WebKit::createItemWithChildren):
(WebKit::createItemRecursive):
(WebKit::createItem):
* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::requestTextExtraction):
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):

Add new test options to allow layout tests to:
- Clip text extraction to the visible bounds of the web view.
- Additionally print rects for each of the extracted items.

* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h:
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.mm:
(WTR::buildDescriptionIgnoringChildren):
(WTR::buildRecursiveDescription):
(WTR::recursiveDescription):

Canonical link: <a href="https://commits.webkit.org/276133@main">https://commits.webkit.org/276133@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f29ec9a98152f0353294f2d37d8f4ffa49ed9a2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43823 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46257 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46467 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39916 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46127 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20276 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36172 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44397 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19908 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37747 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17168 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17431 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38826 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1877 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39998 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39126 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48026 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18843 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15425 "Found 1 new test failure: http/tests/navigation/parsed-iframe-dynamic-form-back-entry.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42981 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20239 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38018 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41673 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9751 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20440 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19863 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -4290,8 +4290,14 @@ - (void)_requestTextExtraction:(CGRect)rectInWebView completionHandler:(void(^)(
+        completionHandler(WebKit::createItem(item, [strongSelf = weakSelf.get()](auto& rectInRootView) {
+#if PLATFORM(IOS_FAMILY)
+            if (RetainPtr contentView = strongSelf ? strongSelf->_contentView : nil)
+                return WebCore::FloatRect { [strongSelf convertRect:rectInRootView fromView:contentView.get()] };
```
I think you can skip `WebCore::FloatRect { }`, right? Since this isn't marked `explicit`.

`WEBCORE_EXPORT FloatRect(const NSRect&);`

whsieh replied:
(Note: I needed to explicitly add a `-> WebCore::FloatRect` for the lambda, otherwise the return value can't be implicitly deduced)

aprotyas commented on Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.h:
```diff
@@ -38,6 +41,8 @@ struct Item;
-RetainPtr<WKTextExtractionItem> createItem(const WebCore::TextExtraction::Item&);
+
+using RootViewToWebViewConverter = Function<WebCore::FloatRect(const WebCore::FloatRect&)>;
+RetainPtr<WKTextExtractionItem> createItem(const WebCore::TextExtraction::Item&, const RootViewToWebViewConverter&);
```
Nit:
```suggestion
RetainPtr<WKTextExtractionItem> createItem(const WebCore::TextExtraction::Item&, RootViewToWebViewConverter&&);
```

aprotyas replied:
Oh, actually, maybe not given how `createItemRecursive` works.

whsieh replied:
Yeah, we could make the first call to `createItem` an rvalue, but then we'll need const ref for the recursive calls.

aprotyas commented on Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.mm:
```diff
@@ -61,8 +62,15 @@ ASCIILiteral description(WKTextExtractionContainer container)
+        if (includeRects == IncludeRects::No)
+            return;
+        auto rect = item.rectInWebView;
+        stream << " rect=["_s << static_cast<int>(rect.origin.x) << ',' << static_cast<int>(rect.origin.y) << ',' << static_cast<int>(rect.size.width) << ',' << static_cast<int>(rect.size.height) << ']';
```
Optional nit: maybe you could `auto rect = roundedIntRect(item.rectInWebView);` and then avoid the bunch of `static_cast`s in the next line?

whsieh replied:
Nice! Much cleaner.

whsieh replied:
Hm... iOS WKTR doesn't link WebCore, so this won't work. I think it's probably cleaner to abstract this out into a helper method to unduplicated some of these calls.

whsieh committed:
[Text Extraction] Retrieved items should be reported in the coordinate space of the web view
https://bugs.webkit.org/show_bug.cgi?id=270994
rdar://124557587

Reviewed by Abrar Rahman Protyasha.

`WKTextExtractionItem`'s rect should be in web view coordinates, rather than root view coordinates.
Fix this by refactoring the text extraction item conversion code, so that we can pass in a function
to convert from root view coordinates to web view coordinates if needed (i.e. on iOS).

Test: fast/text-extraction/text-extraction-when-scrolled.html

* LayoutTests/fast/text-extraction/text-extraction-when-scrolled-expected.txt: Added.
* LayoutTests/fast/text-extraction/text-extraction-when-scrolled.html: Added.

Add a new layout test to exercise the change:
1.  Only the bottom link should be extracted
2.  The y-offset of the extracted link should be closer to 0 than 10000 (roughly the full height of
    the view).
3.  The y-offset of the root should be a large negative number close to 10000.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestTextExtraction):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm:
(WebKit::createWKTextItem):
(WebKit::createItemWithChildren):
(WebKit::createItemRecursive):
(WebKit::createItem):
* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::requestTextExtraction):
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):

Add new test options to allow layout tests to:
- Clip text extraction to the visible bounds of the web view.
- Additionally print rects for each of the extracted items.

* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h:
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.mm:
(WTR::buildDescriptionIgnoringChildren):
(WTR::buildRecursiveDescription):
(WTR::recursiveDescription):

Canonical link: https://commits.webkit.org/276133@main
